lnst/Recipes/ENRT/BondingMixin.py

```diff
@@ -0,0 +1,70 @@
+from lnst.Common.Parameters import (
+    IntParam,
+    StrParam,
+    ChoiceParam,
+)
+from lnst.Devices import RemoteDevice, BondDevice
+from lnst.Controller.Recipe import RecipeError
+
+
+class BondingMixin:
+    """
+    The recipe mixin provides additional recipe parameters to configure the
+    bonding device.
+
+        :param bonding_mode:
+            (mandatory test parameter) the bonding mode to be configured on
+            the bond0 device.
+        :param miimon_value:
+            (mandatory test parameter) the miimon interval to be configured
+            on the bond0 device.
+        :param fail_over_mac:
+            the fail_over_mac mode to be configured on the bond0 device.
+    """
+    bonding_mode = StrParam(mandatory=True)
+    miimon_value = IntParam(mandatory=True)
+    fail_over_mac = ChoiceParam(
+        StrParam, choices={"none", "active", "follow"}, default="none"
+    )
+
+    def create_bond_devices(
+        self, bond_devices_specs: dict[str: dict[str: list[RemoteDevice]]] = {}
+    ) -> None:
+        """
+        The derived class should call:
+        ```
+        self.create_bond_devices(
+            {
+                "host1": {
+                    "bond0": [host1.nic1, host1.nic2]
+                },
+                "host2": {
+                    "bond0": [host2.nic1, host2.nic2]
+                }
+            }
+
+        That would create devices accessible by host1.bond0 and host2.bond0
+        )
+        ```
+        """
+        device_params = dict(
+            mode=self.params.bonding_mode,
+            miimon=self.params.miimon_value,
+        )
+
+        if self.params.bonding_mode in ["active-backup", "1"]:
+            device_params["fail_over_mac"] = self.params.fail_over_mac
+
+        for host_str, spec in bond_devices_specs.items():
+            for bond_dev_name, bonded_devices in spec.items():
+                if len({dev.host for dev in bonded_devices}) > 1:
+                    raise RecipeError(
+                        f"Cannot create bond device with ports coming from different hosts, {bonded_devices}"
```

Comment:
the condition you use counts the number of hosts in "bonded_devices", if there's 2 then yes... the devices are from two different hosts and that doesn't work...

however, you can also provide 2 devices from one host, which is different than the host where the bond device should be created - your condition will count 1, but will pass even when it should fail.

for example:
```
        self.create_bond_devices(
            {
                "host1": {
                    "bond0": [host2.nic1, host2.nic2]
                },
            }
```

nics are from host2, but bond on host1.

Reply:
I'm not referring to the host_str in the spec. I'm just checking if all devices in the bond_devices list are on the same host.

Reply:
I could add your condition with a different exception string if you want.

Reply:
> if all devices in the bond_devices list are on the same host.

yes... but it could be the WRONG host. I know you're not checking `host_str`, i'm saying maybe we should?

but yeah... it's a nitpick like i've said, we can skip this if you want.

Reply:
You're right that with your proposal that would cover both cases. I'll do some testing and eventually update this.

Reply:
Fixed, I've also added check if the `host_str` can be resolved in `self.matched.hosts`

Reply:
looks ok